File: android/GrabDrop/app/src/main/java/com/grabdrop/gesture/GestureClassifier.kt
// /GrabDrop/app/src/main/java/com/grabdrop/gesture/GestureClassifier.kt
package com.grabdrop.gesture

import android.content.Context
import android.util.Log
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import ai.onnxruntime.OnnxTensor
import org.json.JSONObject
import java.nio.FloatBuffer
import kotlin.math.*

/**
 * TCN-based gesture classifier using ONNX Runtime.
 *
 * Features (144 dimensions):
 * - Normalized landmarks (63): relative to wrist, divided by palm size
 * - Velocity (63): frame-to-frame difference
 * - Wrist velocity (3): wrist position change
 * - Finger distances (10): distances between fingertips
 * - Finger angles (5): bending angle for each finger
 */
class GestureClassifier(private val context: Context) {

    companion object {
        private const val TAG = "GestureClassifier"
        private const val MODEL_FILE = "gesture_tcn_pruned_quantized.onnx"
        private const val CONFIG_FILE = "config.json"
    }

    // Model configuration
    private var seqLen: Int = 30
    private var featureDim: Int = 144
    private var rawDim: Int = 63
    private var numLandmarks: Int = 21
    private var classNames: List<String> = listOf("grab", "release", "swipe_up", "swipe_down", "noise")

    // Normalization parameters
    private var normalizeMean: FloatArray = FloatArray(0)
    private var normalizeStd: FloatArray = FloatArray(0)

    // Feature computation constants
    private var pairs: List<Pair<Int, Int>> = listOf()  // Finger tip pairs for distance
    private var fingerChains: List<List<Int>> = listOf()  // Finger joint chains for angle

    // ONNX Runtime
    private var session: OrtSession? = null
    private var env: OrtEnvironment? = null
    var isInitialized = false
        private set

    // Sliding window for sequence (pre-filled with zeros)
    private val window = ArrayDeque<FloatArray>()

    // Track if each frame is real (non-zero) or placeholder
    private val isRealFrame = ArrayDeque<Boolean>()

    // Previous frame for velocity calculation
    private var prevNormLandmarks: FloatArray? = null
    private var prevWrist: FloatArray? = null

    // Landmark indices
    private val WRIST_IDX = 0
    private val MID_FINGER_IDX = 9

    data class ClassificationResult(
        val gesture: String,
        val confidence: Float,
        val classIndex: Int,
        val validFrames: Int  // Number of non-zero frames in window
    )

    // Minimum number of real frames needed to return a result
    private val MIN_VALID_FRAMES = 15

    init {
        try {
            Log.d(TAG, "Initializing GestureClassifier...")
            loadConfig()
            Log.d(TAG, "Config loaded successfully")
            loadModel()
            Log.d(TAG, "Model loaded successfully")
            initializeWindow()
            Log.d(TAG, "Window initialized")
            isInitialized = true
            Log.d(TAG, "GestureClassifier initialized: seqLen=$seqLen, featureDim=$featureDim")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize GestureClassifier: ${e.message}", e)
            isInitialized = false
        }
    }

    private fun loadConfig() {
        try {
            Log.d(TAG, "Loading config from assets: $CONFIG_FILE")

            // Check if file exists
            val assetFiles = context.assets.list("") ?: arrayOf()
            if (!assetFiles.contains(CONFIG_FILE)) {
                throw RuntimeException("Config file not found in assets: $CONFIG_FILE. Available: ${assetFiles.joinToString()}")
            }

            val configJson = context.assets.open(CONFIG_FILE).bufferedReader().use { it.readText() }
            Log.d(TAG, "Config JSON loaded: ${configJson.length} chars")

            val json = JSONObject(configJson)

            seqLen = json.getInt("seq_len")
            featureDim = json.getInt("feature_dim")
            rawDim = json.getInt("raw_dim")
            numLandmarks = json.getInt("num_landmarks")

            // Load class names
            val classNamesArray = json.getJSONArray("class_names")
            classNames = (0 until classNamesArray.length()).map { classNamesArray.getString(it) }

            // Load normalization parameters
            val meanArray = json.getJSONArray("normalize_mean")
            val stdArray = json.getJSONArray("normalize_std")
            normalizeMean = FloatArray(meanArray.length()) { meanArray.getDouble(it).toFloat() }
            normalizeStd = FloatArray(stdArray.length()) { stdArray.getDouble(it).toFloat() }

            // Load pairs for finger distance
            val pairsArray = json.getJSONArray("pairs")
            pairs = (0 until pairsArray.length()).map { i ->
                val pair = pairsArray.getJSONArray(i)
                Pair(pair.getInt(0), pair.getInt(1))
            }

            // Load finger chains for angle calculation
            val chainsArray = json.getJSONArray("finger_chains")
            fingerChains = (0 until chainsArray.length()).map { i ->
                val chain = chainsArray.getJSONArray(i)
                (0 until chain.length()).map { chain.getInt(it) }
            }

            Log.d(TAG, "Config loaded: ${classNames.size} classes, ${pairs.size} pairs, ${fingerChains.size} finger chains")
            Log.d(TAG, "seqLen=$seqLen, featureDim=$featureDim, rawDim=$rawDim, numLandmarks=$numLandmarks")

        } catch (e: Exception) {
            Log.e(TAG, "Error loading config: ${e.message}", e)
            throw e
        }
    }

    private fun loadModel() {
        try {
            Log.d(TAG, "Loading ONNX model from assets: $MODEL_FILE")

            // Check if file exists in assets
            val assetFiles = context.assets.list("") ?: arrayOf()
            if (!assetFiles.contains(MODEL_FILE)) {
                throw RuntimeException("Model file not found in assets: $MODEL_FILE. Available: ${assetFiles.joinToString()}")
            }

            // Create ONNX environment
            env = OrtEnvironment.getEnvironment()
            if (env == null) {
                throw RuntimeException("Failed to create OrtEnvironment")
            }
            Log.d(TAG, "OrtEnvironment created")

            // Load model bytes from assets
            val modelBytes = context.assets.open(MODEL_FILE).use { it.readBytes() }
            Log.d(TAG, "Model bytes loaded: ${modelBytes.size} bytes")

            // Create session options
            val options = OrtSession.SessionOptions()
            options.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)

            // Create session
            session = env!!.createSession(modelBytes, options)
            Log.d(TAG, "ONNX session created: inputNames=${session?.inputNames}, outputNames=${session?.outputNames}")

        } catch (e: Exception) {
            Log.e(TAG, "Error loading ONNX model: ${e.message}", e)
            throw e
        }
    }

    private fun initializeWindow() {
        window.clear()
        isRealFrame.clear()
        // Don't pre-fill - wait for real frames
    }

    /**
     * Reset the sliding window and state.
     */
    fun reset() {
        initializeWindow()
        prevNormLandmarks = null
        prevWrist = null
    }

    /**
     * Add a frame of landmarks and get classification result.
     * @param landmarks Raw landmarks as FloatArray of size 63 (21 points x 3 coords)
     * @return ClassificationResult or null if model not initialized or not enough valid frames
     */
    fun addFrameAndClassify(landmarks: FloatArray?): ClassificationResult? {
        if (!isInitialized || session == null) return null

        // Compute features
        val features = computeFeatures(landmarks)

        // Add to window (sliding, remove oldest if full)
        if (window.size >= seqLen) {
            window.removeFirst()
            isRealFrame.removeFirst()
        }
        window.addLast(features)
        // Mark as real frame if we have valid landmarks
        isRealFrame.addLast(landmarks != null && landmarks.size >= rawDim)

        // Count valid (real) frames in window
        val validFrames = isRealFrame.count { it }
        if (validFrames < MIN_VALID_FRAMES) {
            return null  // Not enough real frames yet
        }

        // Run inference
        return classify(validFrames)
    }

    /**
     * Compute 144-dimensional feature vector from raw landmarks.
     */
    private fun computeFeatures(landmarks: FloatArray?): FloatArray {
        val features = FloatArray(featureDim)

        // Handle null or invalid input
        if (landmarks == null || landmarks.size < rawDim) {
            // Use zeros for invalid input, but keep velocity from previous
            if (prevNormLandmarks != null) {
                // Copy previous velocity (indices 63-125)
                System.arraycopy(prevNormLandmarks!!, 0, features, 63, 63)
            }
            return features
        }

        // Reshape landmarks to (21, 3)
        val lms = Array(numLandmarks) { i ->
            floatArrayOf(
                landmarks[i * 3],
                landmarks[i * 3 + 1],
                landmarks[i * 3 + 2]
            )
        }

        // Extract wrist and mid finger
        val wrist = lms[WRIST_IDX]
        val midFinger = lms[MID_FINGER_IDX]

        // Calculate palm size (distance from wrist to middle finger MCP)
        val palmSize = sqrt(
            (midFinger[0] - wrist[0]).pow(2) +
            (midFinger[1] - wrist[1]).pow(2) +
            (midFinger[2] - wrist[2]).pow(2)
        ).coerceAtLeast(1e-6f)

        // 1. Normalized landmarks (63 dims): relative to wrist, divided by palm size
        val normLandmarks = FloatArray(rawDim)
        for (i in 0 until numLandmarks) {
            normLandmarks[i * 3] = (lms[i][0] - wrist[0]) / palmSize
            normLandmarks[i * 3 + 1] = (lms[i][1] - wrist[1]) / palmSize
            normLandmarks[i * 3 + 2] = (lms[i][2] - wrist[2]) / palmSize
        }
        System.arraycopy(normLandmarks, 0, features, 0, rawDim)

        // 2. Velocity (63 dims): frame-to-frame difference of normalized landmarks
        val velocity = if (prevNormLandmarks != null) {
            FloatArray(rawDim) { i ->
                normLandmarks[i] - prevNormLandmarks!![i]
            }
        } else {
            FloatArray(rawDim) { 0f }
        }
        System.arraycopy(velocity, 0, features, rawDim, rawDim)

        // 3. Wrist velocity (3 dims)
        val wristVel = if (prevWrist != null) {
            floatArrayOf(
                wrist[0] - prevWrist!![0],
                wrist[1] - prevWrist!![1],
                wrist[2] - prevWrist!![2]
            )
        } else {
            floatArrayOf(0f, 0f, 0f)
        }
        features[rawDim * 2] = wristVel[0]
        features[rawDim * 2 + 1] = wristVel[1]
        features[rawDim * 2 + 2] = wristVel[2]

        // 4. Finger distances (10 dims): distances between fingertip pairs
        val offset = rawDim * 2 + 3
        for ((idx, pair) in pairs.withIndex()) {
            val (i, j) = pair
            val dx = normLandmarks[i * 3] - normLandmarks[j * 3]
            val dy = normLandmarks[i * 3 + 1] - normLandmarks[j * 3 + 1]
            val dz = normLandmarks[i * 3 + 2] - normLandmarks[j * 3 + 2]
            features[offset + idx] = sqrt(dx * dx + dy * dy + dz * dz)
        }

        // 5. Finger angles (5 dims): bending angle for each finger
        val angleOffset = offset + pairs.size
        for ((idx, chain) in fingerChains.withIndex()) {
            if (chain.size < 3) continue

            // v1: from chain[0] to chain[1]
            // v2: from chain[1] to chain[last] (fingertip)
            val v1 = floatArrayOf(
                lms[chain[1]][0] - lms[chain[0]][0],
                lms[chain[1]][1] - lms[chain[0]][1],
                lms[chain[1]][2] - lms[chain[0]][2]
            )
            val lastIdx = chain[chain.size - 1]
            val v2 = floatArrayOf(
                lms[lastIdx][0] - lms[chain[1]][0],
                lms[lastIdx][1] - lms[chain[1]][1],
                lms[lastIdx][2] - lms[chain[1]][2]
            )

            val n1 = sqrt(v1[0] * v1[0] + v1[1] * v1[1] + v1[2] * v1[2]) + 1e-8f
            val n2 = sqrt(v2[0] * v2[0] + v2[1] * v2[1] + v2[2] * v2[2]) + 1e-8f

            val cosAngle = ((v1[0] / n1) * (v2[0] / n2) +
                           (v1[1] / n1) * (v2[1] / n2) +
                           (v1[2] / n1) * (v2[2] / n2)).coerceIn(-1f, 1f)

            features[angleOffset + idx] = acos(cosAngle)
        }

        // Save current state for next frame
        prevNormLandmarks = normLandmarks.copyOf()
        prevWrist = wrist.copyOf()

        // Debug: log some feature values
        if (Log.isLoggable(TAG, Log.DEBUG)) {
            // Log finger angles (last 5 dims) - should be different for FIST vs PALM
            val angles = FloatArray(5) { features[angleOffset + it] }
            Log.d(TAG, "Features: angles=[${angles.map { "%.2f".format(it) }.joinToString(", ")}] " +
                    "palmSize=${String.format("%.4f", palmSize)}")
        }

        return features
    }

    /**
     * Run inference on the current window.
     * @param validFrames Number of real (non-placeholder) frames in window
     */
    private fun classify(validFrames: Int): ClassificationResult? {
        val session = this.session ?: return null
        val env = this.env ?: return null

        try {
            // Prepare input: shape (1, featureDim, seqLen)
            // Real frames at the BEGINNING (time 0), zeros at the END
            val inputArray = Array(1) { Array(featureDim) { FloatArray(seqLen) } }
            val windowList = window.toList()
            val numFrames = minOf(windowList.size, seqLen)

            for (t in 0 until numFrames) {
                val frame = windowList[t]
                for (d in 0 until minOf(frame.size, featureDim)) {
                    // Normalize
                    val normalized = (frame[d] - normalizeMean[d]) / (normalizeStd[d] + 1e-8f)
                    inputArray[0][d][t] = normalized  // Place real frames at the beginning
                }
            }

            // Create ONNX tensor
            val inputName = session.inputNames.iterator().next()
            val inputTensor = OnnxTensor.createTensor(env, inputArray)

            // Run inference
            val output = session.run(mapOf(inputName to inputTensor))

            // Get output
            val outputTensor = output[0] as OnnxTensor
            Log.d(TAG, "Output shape: ${outputTensor.info.shape.joinToString()}")
            val outputBuffer = outputTensor.floatBuffer

            // Read output values from buffer
            val outputArray = FloatArray(classNames.size)
            outputBuffer.rewind()
            for (i in outputArray.indices) {
                outputArray[i] = outputBuffer.get()
            }

            // Debug: log raw outputs
            val outputStr = outputArray.mapIndexed { i, v ->
                "${classNames[i]}=${String.format("%.2f", v)}"
            }.joinToString(" ")
            Log.d(TAG, "Raw output: [$outputStr]")

            // Find max class
            var maxIdx = 0
            var maxVal = outputArray[0]
            for (i in 1 until classNames.size) {
                if (outputArray[i] > maxVal) {
                    maxVal = outputArray[i]
                    maxIdx = i
                }
            }

            // Apply softmax for confidence
            val maxLogit = outputArray.maxOrNull() ?: 0f
            var sumExp = 0f
            val probs = FloatArray(classNames.size) { i ->
                val exp = kotlin.math.exp(outputArray[i] - maxLogit)
                sumExp += exp
                exp
            }
            val confidence = probs[maxIdx] / sumExp

            inputTensor.close()
            output.close()

            return ClassificationResult(
                gesture = classNames[maxIdx],
                confidence = confidence,
                classIndex = maxIdx,
                validFrames = validFrames
            )
        } catch (e: Exception) {
            Log.e(TAG, "Classification error: ${e.message}", e)
            return null
        }
    }

    fun close() {
        try {
            session?.close()
            session = null
            env = null
        } catch (e: Exception) {
            Log.e(TAG, "Error closing classifier", e)
        }
    }
}

File: desktop/gesture_classifier.py
#!/usr/bin/python

#*================================================================
#*   Copyright (C) 2026 XUranus All rights reserved.
#*
#*   File:         gesture_classifier.py
#*   Author:       XUranus
#*   Date:         2026-03-24
#*   Description:  TCN-based gesture classifier using ONNX Runtime.
#*                 Python port of the Android GestureClassifier.kt.
#*
#================================================================*/

"""
TCN-based gesture classifier using ONNX Runtime.

Features (144 dimensions):
- Normalized landmarks (63): relative to wrist, divided by palm size
- Velocity (63): frame-to-frame difference
- Wrist velocity (3): wrist position change
- Finger distances (10): distances between fingertip pairs
- Finger angles (5): bending angle for each finger
"""

import json
import logging
import math
from collections import deque
from dataclasses import dataclass
from typing import List, Optional, Tuple

import numpy as np

import config

logger = logging.getLogger("GestureClassifier")


@dataclass
class ClassificationResult:
    gesture: str
    confidence: float
    class_index: int
    valid_frames: int  # Number of real (non-zero) frames in window


class GestureClassifier:
    WRIST_IDX = 0
    MID_FINGER_IDX = 9  # middle finger MCP

    def __init__(self):
        self.is_initialized = False

        # Model configuration (loaded from config.json)
        self.seq_len: int = 30
        self.feature_dim: int = 144
        self.raw_dim: int = 63
        self.num_landmarks: int = 21
        self.class_names: List[str] = ["grab", "release", "swipe_up", "swipe_down", "noise"]

        # Normalization parameters
        self.normalize_mean: np.ndarray = np.zeros(0)
        self.normalize_std: np.ndarray = np.zeros(0)

        # Feature computation
        self.pairs: List[Tuple[int, int]] = []
        self.finger_chains: List[List[int]] = []

        # ONNX Runtime session
        self.session = None

        # Sliding window
        self.window: deque = deque()
        self.is_real_frame: deque = deque()

        # Previous frame state for velocity
        self._prev_norm_landmarks: Optional[np.ndarray] = None
        self._prev_wrist: Optional[np.ndarray] = None

        # Minimum real frames before classification
        self.min_valid_frames = 15

        try:
            logger.info("Initializing GestureClassifier...")
            self._load_config()
            self._load_model()
            self._init_window()
            self.is_initialized = True
            logger.info(
                f"GestureClassifier initialized: "
                f"seq_len={self.seq_len}, feature_dim={self.feature_dim}, "
                f"classes={self.class_names}"
            )
        except Exception as e:
            logger.error(f"Failed to initialize GestureClassifier: {e}")
            self.is_initialized = False

    def _load_config(self):
        config_path = config.TCN_CONFIG_PATH
        logger.info(f"Loading TCN config from {config_path}")

        with open(config_path, "r") as f:
            cfg = json.load(f)

        self.seq_len = cfg["seq_len"]
        self.feature_dim = cfg["feature_dim"]
        self.raw_dim = cfg["raw_dim"]
        self.num_landmarks = cfg["num_landmarks"]
        self.class_names = cfg["class_names"]

        self.normalize_mean = np.array(cfg["normalize_mean"], dtype=np.float32)
        self.normalize_std = np.array(cfg["normalize_std"], dtype=np.float32)

        self.pairs = [tuple(p) for p in cfg["pairs"]]
        self.finger_chains = cfg["finger_chains"]

        logger.info(
            f"Config: {len(self.class_names)} classes, "
            f"{len(self.pairs)} pairs, {len(self.finger_chains)} chains"
        )

    def _load_model(self):
        import onnxruntime as ort

        model_path = config.TCN_MODEL_PATH
        logger.info(f"Loading ONNX model from {model_path}")

        opts = ort.SessionOptions()
        opts.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL

        self.session = ort.InferenceSession(model_path, opts)

        input_info = self.session.get_inputs()[0]
        output_info = self.session.get_outputs()[0]
        logger.info(
            f"Model loaded: input={input_info.name} {input_info.shape}, "
            f"output={output_info.name} {output_info.shape}"
        )

    def _init_window(self):
        self.window.clear()
        self.is_real_frame.clear()

    def reset(self):
        """Reset sliding window and velocity state."""
        self._init_window()
        self._prev_norm_landmarks = None
        self._prev_wrist = None

    def add_frame_and_classify(
        self, landmarks: Optional[np.ndarray]
    ) -> Optional[ClassificationResult]:
        """
        Add a frame of landmarks and get classification result.

        Args:
            landmarks: Raw landmarks as ndarray of shape (63,) — 21 points x 3 coords.
                       Pass None if no hand detected.

        Returns:
            ClassificationResult or None if not enough frames yet.
        """
        if not self.is_initialized or self.session is None:
            return None

        features = self._compute_features(landmarks)

        # Sliding window
        if len(self.window) >= self.seq_len:
            self.window.popleft()
            self.is_real_frame.popleft()
        self.window.append(features)
        self.is_real_frame.append(
            landmarks is not None and len(landmarks) >= self.raw_dim
        )

        valid_frames = sum(self.is_real_frame)
        if valid_frames < self.min_valid_frames:
            return None

        return self._classify(valid_frames)

    def _compute_features(self, landmarks: Optional[np.ndarray]) -> np.ndarray:
        """Compute 144-dimensional feature vector from raw landmarks."""
        features = np.zeros(self.feature_dim, dtype=np.float32)

        if landmarks is None or len(landmarks) < self.raw_dim:
            return features

        # Reshape to (21, 3)
        lms = landmarks.reshape(self.num_landmarks, 3)

        wrist = lms[self.WRIST_IDX]
        mid_finger = lms[self.MID_FINGER_IDX]

        # Palm size
        palm_size = max(np.linalg.norm(mid_finger - wrist), 1e-6)

        # 1. Normalized landmarks (63 dims)
        norm_landmarks = ((lms - wrist) / palm_size).flatten()
        features[0:self.raw_dim] = norm_landmarks

        # 2. Velocity (63 dims)
        if self._prev_norm_landmarks is not None:
            velocity = norm_landmarks - self._prev_norm_landmarks
        else:
            velocity = np.zeros(self.raw_dim, dtype=np.float32)
        features[self.raw_dim : self.raw_dim * 2] = velocity

        # 3. Wrist velocity (3 dims)
        offset = self.raw_dim * 2
        if self._prev_wrist is not None:
            features[offset : offset + 3] = wrist - self._prev_wrist
        # else: zeros already

        # 4. Finger distances (10 dims)
        offset += 3
        for idx, (i, j) in enumerate(self.pairs):
            d = np.linalg.norm(
                norm_landmarks[i * 3 : i * 3 + 3]
                - norm_landmarks[j * 3 : j * 3 + 3]
            )
            features[offset + idx] = d

        # 5. Finger angles (5 dims)
        offset += len(self.pairs)
        for idx, chain in enumerate(self.finger_chains):
            if len(chain) < 3:
                continue
            v1 = lms[chain[1]] - lms[chain[0]]
            v2 = lms[chain[-1]] - lms[chain[1]]

            n1 = np.linalg.norm(v1) + 1e-8
            n2 = np.linalg.norm(v2) + 1e-8

            cos_angle = np.clip(np.dot(v1 / n1, v2 / n2), -1.0, 1.0)
            features[offset + idx] = math.acos(cos_angle)

        # Save state for next frame
        self._prev_norm_landmarks = norm_landmarks.copy()
        self._prev_wrist = wrist.copy()

        return features

    def _classify(self, valid_frames: int) -> Optional[ClassificationResult]:
        """Run inference on the current window."""
        if self.session is None:
            return None

        try:
            # Build input: shape (1, feature_dim, seq_len)
            input_array = np.zeros(
                (1, self.feature_dim, self.seq_len), dtype=np.float32
            )

            window_list = list(self.window)
            num_frames = min(len(window_list), self.seq_len)

            for t in range(num_frames):
                frame = window_list[t]
                for d in range(min(len(frame), self.feature_dim)):
                    normalized = (frame[d] - self.normalize_mean[d]) / (
                        self.normalize_std[d] + 1e-8
                    )
                    input_array[0, d, t] = normalized

            input_name = self.session.get_inputs()[0].name
            outputs = self.session.run(None, {input_name: input_array})

            logits = outputs[0].flatten()

            # Softmax for confidence
            max_logit = logits.max()
            exp_logits = np.exp(logits - max_logit)
            probs = exp_logits / exp_logits.sum()

            max_idx = int(np.argmax(probs))
            confidence = float(probs[max_idx])
            gesture = self.class_names[max_idx]

            return ClassificationResult(
                gesture=gesture,
                confidence=confidence,
                class_index=max_idx,
                valid_frames=valid_frames,
            )

        except Exception as e:
            logger.error(f"Classification error: {e}")
            return None

    def close(self):
        """Release ONNX session."""
        self.session = None
        logger.info("GestureClassifier closed")
